File: library/src/main/java/su/levenetc/android/textsurface/animations/CamRot.java
package su.levenetc.android.textsurface.animations;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.graphics.PointF;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import su.levenetc.android.textsurface.utils.AnimatorEndListener;
import su.levenetc.android.textsurface.interfaces.ICameraAnimation;
import su.levenetc.android.textsurface.interfaces.IEndListener;
import su.levenetc.android.textsurface.contants.Pivot;
import su.levenetc.android.textsurface.SurfaceCamera;
import su.levenetc.android.textsurface.Text;
import su.levenetc.android.textsurface.TextSurface;

/**
 * Created by Eugene Levenetc.
 */
public class CamRot implements ICameraAnimation, ValueAnimator.AnimatorUpdateListener {

	public float rotationDelta;
	private float pivotX;
	private float pivotY;
	private int duration;
	private TextSurface textSurface;
	private SurfaceCamera camera;
	private Text textPivot;
	private int pivotAlign;
	private ObjectAnimator animator;

	public CamRot(int duration, float rotationDelta) {
		this.duration = duration;
		this.rotationDelta = rotationDelta;
	}

	public CamRot(int duration, float rotationDelta, float pivotX, float pivotY) {
		this.duration = duration;
		this.rotationDelta = rotationDelta;
		this.pivotX = pivotX;
		this.pivotY = pivotY;
	}

	public CamRot(int duration, float rotationDelta, Text textPivot, int pivotAlign) {
		this.duration = duration;
		this.rotationDelta = rotationDelta;
		this.textPivot = textPivot;
		this.pivotAlign = pivotAlign;
	}

	@Override public void onStart() {

	}

	@Override public void start(@Nullable final IEndListener listener) {

		setPivot();
		float from = camera.getRotation();
		float to = camera.getRotation() + rotationDelta;
		animator = ObjectAnimator.ofFloat(camera, "rotation", from, to);
		animator.setDuration(duration);
		animator.addUpdateListener(this);
		animator.addListener(new AnimatorEndListener() {
			@Override public void onAnimationEnd(Animator animation) {
				if (listener != null) listener.onAnimationEnd(CamRot.this);
			}
		});

		animator.start();
	}

	private PointF tmpPoint = new PointF();

	private void setPivot() {

		PointF rotationPivot = camera.getRotationPivot();

		if (textPivot != null) {
			if ((pivotAlign & Pivot.LEFT) == Pivot.LEFT) {
				pivotX = textPivot.getX(textSurface);
			} else if ((pivotAlign & Pivot.RIGHT) == Pivot.RIGHT) {
				pivotX = textPivot.getX(textSurface) + textPivot.getWidth();
			}

			if ((pivotAlign & Pivot.BOTTOM) == Pivot.BOTTOM) {
				pivotY = textPivot.getY(textSurface);
			} else if ((pivotAlign & Pivot.TOP) == Pivot.TOP) {
				pivotY = textPivot.getY(textSurface) - textPivot.getHeight();
			}

			if ((pivotAlign & Pivot.CENTER) == Pivot.CENTER) {
				pivotX = textPivot.getX(textSurface) + textPivot.getWidth() / 2;
				pivotY = textPivot.getY(textSurface) - textPivot.getHeight() / 2;
			}

		}

		tmpPoint.set(pivotX, pivotY);
//		rotatePoint(tmpPoint, camera.getRotationPivot(), camera.getRotation());
		rotatePoint2(tmpPoint, camera.getRotationPivot(), camera.getRotation());
		rotationPivot.set(tmpPoint.x, tmpPoint.y);
	}

	private static void rotatePoint2(PointF point, PointF center, float angle) {

		float x = angle;
		float centerX = center.x;
		float centerY = center.y;
		float point2x = point.x;
		float point2y = point.y;

		double newX = centerX + (point2x - centerX) * Math.cos(x) - (point2y - centerY) * Math.sin(x);
		double newY = centerY + (point2x - centerX) * Math.sin(x) + (point2y - centerY) * Math.cos(x);
		point.set((float) newX, (float) newY);
	}

	private static void rotatePoint(PointF point, PointF center, float angle) {

		if (angle == 0) return;

		double x1 = point.x - center.x;
		double y1 = point.y - center.y;

		//APPLY ROTATION
		double xTemp = x1;
		x1 = x1 * Math.cos(angle) - y1 * Math.sin(angle);
		y1 = xTemp * Math.sin(angle) + y1 * Math.cos(angle);

		//TRANSLATE BACK
		point.x = (int) Math.ceil(x1) + center.x;
		point.y = (int) Math.ceil(y1) + center.y;
	}

	@Override public void setTextSurface(@NonNull TextSurface textSurface) {
		this.textSurface = textSurface;
	}

	@Override public long getDuration() {
		return duration;
	}

	@Override public void cancel() {
		if (animator != null && animator.isRunning()) {
			animator.cancel();
			animator = null;
		}
	}

	@Override public void onAnimationUpdate(ValueAnimator animation) {
		textSurface.invalidate();
	}

	@Override public void setCamera(SurfaceCamera camera) {
		this.camera = camera;
	}
}

File: library/src/main/java/su/levenetc/android/textsurface/animations/CamRot.kt
package su.levenetc.android.textsurface.animations

import android.animation.Animator
import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.animation.ValueAnimator.AnimatorUpdateListener
import android.annotation.SuppressLint
import android.graphics.PointF
import su.levenetc.android.textsurface.SurfaceCamera
import su.levenetc.android.textsurface.Text
import su.levenetc.android.textsurface.TextSurface
import su.levenetc.android.textsurface.contants.Pivot
import su.levenetc.android.textsurface.interfaces.ICameraAnimation
import su.levenetc.android.textsurface.interfaces.IEndListener
import su.levenetc.android.textsurface.utils.AnimatorEndListener

/**
 * Created by Eugene Levenetc.
 */
class CamRot : ICameraAnimation, AnimatorUpdateListener {
    var rotationDelta: Float
    private var pivotX = 0f
    private var pivotY = 0f
    override var duration: Long
    override lateinit var textSurface: TextSurface
    private var camera: SurfaceCamera? = null
    private var textPivot: Text? = null
    private var pivotAlign = 0
    private var animator: ObjectAnimator? = null

    constructor(duration: Long, rotationDelta: Float) {
        this.duration = duration
        this.rotationDelta = rotationDelta
    }

    constructor(duration: Long, rotationDelta: Float, pivotX: Float, pivotY: Float) {
        this.duration = duration
        this.rotationDelta = rotationDelta
        this.pivotX = pivotX
        this.pivotY = pivotY
    }

    constructor(duration: Long, rotationDelta: Float, textPivot: Text?, pivotAlign: Int) {
        this.duration = duration
        this.rotationDelta = rotationDelta
        this.textPivot = textPivot
        this.pivotAlign = pivotAlign
    }

    override fun onStart() {}

    @SuppressLint("ObjectAnimatorBinding")
    override fun start(listener: IEndListener) {
        setPivot()
        val from = camera!!.rotation
        val to = camera!!.rotation + rotationDelta
        animator = ObjectAnimator.ofFloat(camera, "rotation", from, to)
        animator!!.setDuration(duration.toLong())
        animator!!.addUpdateListener(this)
        animator!!.addListener(object : AnimatorEndListener() {
            override fun onAnimationEnd(animation: Animator) {
                listener.onAnimationEnd(this@CamRot)
            }
        })
        animator!!.start()
    }

    private val tmpPoint = PointF()
    private fun setPivot() {
        val rotationPivot = camera!!.rotationPivot
        if (textPivot != null) {
            if (pivotAlign and Pivot.LEFT == Pivot.LEFT) {
                pivotX = textPivot!!.getX(textSurface)
            } else if (pivotAlign and Pivot.RIGHT == Pivot.RIGHT) {
                pivotX = textPivot!!.getX(textSurface) + textPivot!!.width
            }
            if (pivotAlign and Pivot.BOTTOM == Pivot.BOTTOM) {
                pivotY = textPivot!!.getY(textSurface)
            } else if (pivotAlign and Pivot.TOP == Pivot.TOP) {
                pivotY = textPivot!!.getY(textSurface) - textPivot!!.height
            }
            if (pivotAlign and Pivot.CENTER == Pivot.CENTER) {
                pivotX = textPivot!!.getX(textSurface) + textPivot!!.width / 2
                pivotY = textPivot!!.getY(textSurface) - textPivot!!.height / 2
            }
        }
        tmpPoint[pivotX] = pivotY
        //		rotatePoint(tmpPoint, camera.getRotationPivot(), camera.getRotation());
        rotatePoint2(tmpPoint, camera!!.rotationPivot, camera!!.rotation)
        rotationPivot[tmpPoint.x] = tmpPoint.y
    }

    override fun cancel() {
        if (animator != null && animator!!.isRunning) {
            animator!!.cancel()
            animator = null
        }
    }

    override fun onAnimationUpdate(animation: ValueAnimator) {
        textSurface!!.invalidate()
    }

    override fun setCamera(camera: SurfaceCamera?) {
        this.camera = camera
    }

    companion object {
        private fun rotatePoint2(point: PointF, center: PointF, angle: Float) {
            val centerX = center.x
            val centerY = center.y
            val point2x = point.x
            val point2y = point.y
            val newX = centerX + (point2x - centerX) * Math.cos(angle.toDouble()) - (point2y - centerY) * Math.sin(angle.toDouble())
            val newY = centerY + (point2x - centerX) * Math.sin(angle.toDouble()) + (point2y - centerY) * Math.cos(angle.toDouble())
            point[newX.toFloat()] = newY.toFloat()
        }

        private fun rotatePoint(point: PointF, center: PointF, angle: Float) {
            if (angle == 0f) return
            var x1 = (point.x - center.x).toDouble()
            var y1 = (point.y - center.y).toDouble()

            //APPLY ROTATION
            val xTemp = x1
            x1 = x1 * Math.cos(angle.toDouble()) - y1 * Math.sin(angle.toDouble())
            y1 = xTemp * Math.sin(angle.toDouble()) + y1 * Math.cos(angle.toDouble())

            //TRANSLATE BACK
            point.x = Math.ceil(x1).toInt() + center.x
            point.y = Math.ceil(y1).toInt() + center.y
        }
    }
}